File: 47-permutations-ii/47-permutations-ii.cpp
class Solution {
public:
    
    
vector<vector<int>> res;

void dfs(vector<int>& nums , vector<int> perm , map<int , int> hash){

	if((int)perm.size() == (int)nums.size()){
		res.push_back(perm);
		return ;
	}

	for (auto i : hash){
		if (i.second > 0 ){
			perm.push_back(i.first);
			hash[i.first] --;

			dfs(nums , perm , hash);

			hash[i.first] ++;
			perm.pop_back();

		}
	}

}


    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        
	vector<int> perm;
	map<int , int> hash;
	for (int i : nums){
		hash[i] ++;
	}

	dfs(nums , perm , hash);

	return res;

    }
};

File: 47-permutations-ii/47-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
            perm = []
            res = []
            count = {i: 0 for i in nums}
            for i in nums:
                count[i] +=1

            def dfs():
                #base case
                if len(perm) == len(nums):
                    res.append(perm.copy())
                    return

                for i in count :
                    if count[i] > 0 :
                        perm.append(i)
                        count[i] -= 1
                        dfs()
                        count[i] +=1
                        perm.pop()
            dfs()
            return res

        